### PR TITLE
[Core] Adding `FromConnectivityGenerateEntities`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Adding `FromConnectivityGenerateEntities` methods in order to generate conditions/elements from connectivities provided with `std::vector<std::vector<std::size_t>>`. The method works in parallel.

**🆕 Changelog**

- [Adding `FromConnectivityGenerateEntities`](https://github.com/KratosMultiphysics/Kratos/commit/839cdfe6bc3186c67e58e4e07fad98e59d7e0b11)
- [Expose to python](https://github.com/KratosMultiphysics/Kratos/commit/c216b26c8cde70772722846e151b93d1584fc05d)
- [Adding tests](https://github.com/KratosMultiphysics/Kratos/commit/1ee425a00c46df3a4b532cc8d26be001c1ffb5fa)

### CODE REVIEW

**Comment by sunethwarna** [comment]
Hi, small suggestion, woudn't it be better to have another method to fill `rThisNodes` container as well from another container (element or condition). This way, someone can change `SurfaceCondition3D3N` to an `Element3D3N` if required.

**Comment by loumalouomega** [comment]
> Hi, this is very useful for me too (I have an implementation which does the same)... a small suggestion, woudn't it be better to have another method to fill `rThisNodes` container as well from another container (element or condition). This way, someone can change `SurfaceCondition3D3N` to an `Element3D3N` if required.

I don't understand what you mean

**Comment by sunethwarna** [comment]
```python
def AddNodes(model_part, container):
     for enttiy in container:
             for node in entity.GetNodes():
                   model_part.AddNode(node)
```

Basically the above, but in c++, with more optimized code (The above is naive). And filling the local_mesh, ghose mesh correctly.

File: kratos/python/add_other_utilities_to_python.cpp
Comment (sunethwarna) [inline kratos/python/add_other_utilities_to_python.cpp]: I would add `py::arg` to the arguments.

**Comment by loumalouomega** [comment]
> ```python
> def AddNodes(model_part, container):
>      for enttiy in container:
>              for node in entity.GetNodes():
>                    model_part.AddNode(node)
> ```
> 
> Basically the above, but in c++, with more optimized code (The above is naive). And filling the local_mesh, ghose mesh correctly.

That's already done in the auxiliary_model_part_utilities

File: kratos/python/add_other_utilities_to_python.cpp
Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp]: Meaning?

Comment (sunethwarna) [reply kratos/python/add_other_utilities_to_python.cpp]:
```suggestion
        .def("FromConnectivityGenerateElements", [](
            const std::string& rEntityName,
            const std::vector<std::vector<std::size_t>>& rEntitiesConnectivities,
            ModelPart::NodesContainerType& rThisNodes,
            ModelPart::ElementsContainerType& rThisElements,
            const Properties::Pointer pProperties) {
                ModelPartUtils::FromConnectivityGenerateEntities<Element>(rEntityName, rEntitiesConnectivities, rThisNodes, rThisElements, pProperties);
            })
        .def("FromConnectivityGenerateConditions", [](
            const std::string& rEntityName,
            const std::vector<std::vector<std::size_t>>& rEntitiesConnectivities,
            ModelPart::NodesContainerType& rThisNodes,
            ModelPart::ConditionsContainerType& rThisConditions,
            const Properties::Pointer pProperties) {
                ModelPartUtils::FromConnectivityGenerateEntities<Condition>(rEntityName, rEntitiesConnectivities, rThisNodes, rThisConditions, pProperties);
            }, py::arg("condition_name"), py::arg("condition_connectivities_list"), ...)
    ;
```

Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp]: Ah, ok

**Comment by sunethwarna** [comment]
> That's already done in the auxiliary_model_part_utilities

Could you let me know which one is it? The deepcopy copies everything, but for your this method, you only need the nodes of the interested container.

**Comment by loumalouomega** [comment]
> > That's already done in the auxiliary_model_part_utilities
> 
> Could you let me know which one is it? The deepcopy copies everything, but for your this method, you only need the nodes of the interested container.

Not merged yet :P https://github.com/KratosMultiphysics/Kratos/pull/7577 @pooyan-dadvand

**Comment by rubenzorrilla** [comment]
Two comments:
1. Maybe a bit picky but I'd call it `GenerateEntitiesFromConnectivities`.
2. Would be great to make a modeller on top of this (of course not for this PR).

**Comment by loumalouomega** [comment]
> Two comments:
> 
>     1. Maybe a bit picky but I'd call it `GenerateEntitiesFromConnectivities`.

👍

>     2. Would be great to make a modeller on top of this (of course not for this PR).

👍. Can you define the input ?

**Comment by rubenzorrilla** [comment]
> > ```
> > 2. Would be great to make a modeller on top of this (of course not for this PR).
> > ```
> 
> 👍. Can you define the input ?

I was thinking on giving a directory from which we can parse a `Numpy` or `Pandas` array. I think that something of this sort could be quite handy.

**Comment by loumalouomega** [comment]
> > > ```
> > > 2. Would be great to make a modeller on top of this (of course not for this PR).
> > > ```
> > 
> > 
> > 👍. Can you define the input ?
> 
> I was thinking on giving a directory from which we can parse a `Numpy` or `Pandas` array. I think that something of this sort could be quite handy.
> 
> EDIT: (just thinking out loud)

I don't get you. So in the parameters we provide a folder where we can find a file with the connectivities?

**Comment by sunethwarna** [comment]
> Not merged yet :P https://github.com/KratosMultiphysics/Kratos/pull/7577 @pooyan-dadvand

This is exactly what I needed, now I have done it in the OptApp....  Can we make it to the core @pooyan-dadvand ?

**Comment by rubenzorrilla** [comment]
> > > > ```
> > > > 2. Would be great to make a modeller on top of this (of course not for this PR).
> > > > ```
> > > 
> > > 
> > > 👍. Can you define the input ?
> > 
> > 
> > I was thinking on giving a directory from which we can parse a `Numpy` or `Pandas` array. I think that something of this sort could be quite handy.
> > EDIT: (just thinking out loud)
> 
> I don't get you. So in the parameters we provide a folder where we can find a file with the connectivities?

Exactly this.

**Comment by loumalouomega** [comment]
> > > > > ```
> > > > > 2. Would be great to make a modeller on top of this (of course not for this PR).
> > > > > ```
> > > > 
> > > > 
> > > > 👍. Can you define the input ?
> > > 
> > > 
> > > I was thinking on giving a directory from which we can parse a `Numpy` or `Pandas` array. I think that something of this sort could be quite handy.
> > > EDIT: (just thinking out loud)
> > 
> > 
> > I don't get you. So in the parameters we provide a folder where we can find a file with the connectivities?
> 
> Exactly this.

Huuum, so a dummer mdpa with just connectivities..., I don't know how this cannot be problematic as ids of the nodes may be defined in a inconcistent manner related to existing nodes

**Comment by loumalouomega** [comment]
@roigcarlo the CI of Linux are failing after the merge of the new compilation system. Looks like neither the swimmingDEM is failing or segmentation fault when launching tests.

**Comment by loumalouomega** [comment]
> @roigcarlo the CI of Linux are failing after the merge of the new compilation system. Looks like neither the swimmingDEM is failing or segmentation fault when launching tests.

I see https://github.com/KratosMultiphysics/Kratos/pull/12206

**Comment by roigcarlo** [comment]
Yep sorry, I think its fixed now, lets see if it pases the CI

File: kratos/python/add_other_utilities_to_python.cpp
Comment (philbucher) [inline kratos/python/add_other_utilities_to_python.cpp]: Its a class, not a namespace, this is wrong

**Review by philbucher — COMMENTED** [review]
Conceptually, this should be used by the ConnPreserveModeller no?

File: kratos/python/add_other_utilities_to_python.cpp
Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp]: Okay

File: kratos/utilities/model_part_utils.h
Comment (pooyan-dadvand) [inline kratos/utilities/model_part_utils.h]: Is the comment correct?

**Review by pooyan-dadvand — COMMENTED** [review]
Just a minor comment

**Comment by pooyan-dadvand** [comment]
I think a python test for the interface would be very welcomed!

File: kratos/utilities/model_part_utils.h
Comment (loumalouomega) [reply kratos/utilities/model_part_utils.h]: Yes, `reserve` works with the destiny capacity.

**Comment by loumalouomega** [comment]
> I think a python test for the interface would be very welcomed!

Done

**Comment by pooyan-dadvand** [comment]
For some reason the MPMApplication test is failing....

**Comment by loumalouomega** [comment]
> For some reason the MPMApplication test is failing....

That test fails randomly...

File: kratos/utilities/model_part_utils.h
Comment (pooyan-dadvand) [reply kratos/utilities/model_part_utils.h]: But you are not using the first entity as the one you want to clone. You are getting it from the input.

Comment (loumalouomega) [reply kratos/utilities/model_part_utils.h]: The vector may be empty, this way you ensure you are using the element you are interested. If you want I can add a method with a prototype

Comment (loumalouomega) [reply kratos/utilities/model_part_utils.h]: Ah, you mean the comment, it is a leftover froma  previous iteration, cleaning

Comment (loumalouomega) [reply kratos/utilities/model_part_utils.h]: Done

**Comment by loumalouomega** [comment]
Can anyone approve?, @pooyan-dadvand agrees after last change.

**Comment by loumalouomega** [comment]
@sunethwarna

**Comment by loumalouomega** [comment]
There was a conflict, I fixed it, can anyone approve?

**Review by pooyan-dadvand — APPROVED** [review]

**Pull request merged**
